pandas/tests/groupby/test_apply.py

```diff
@@ -827,3 +827,11 @@ def test_apply_index_has_complex_internals(index):
     df = DataFrame({"group": [1, 1, 2], "value": [0, 1, 0]}, index=index)
     result = df.groupby("group").apply(lambda x: x)
     tm.assert_frame_equal(result, df)
+
+
+def test_apply_function_returns_list():
+    # GH 31441
+    df = pd.DataFrame(["A", "A", "B", "B"], columns=["groups"])
+    result = df.groupby("groups").apply(lambda x: x.index.to_list())
```

Comment:
Might be good to add tests for other return types as well (e.g., set or tuple)

Reply:
Yes, good suggestion, thanks!